src/DOMWorld.ts

```diff
@@ -430,18 +436,24 @@ class WaitTask {
   _timeoutTimer?: NodeJS.Timeout;
   _terminated = false;
 
-  constructor(domWorld: DOMWorld, predicateBody: Function | string, title: string, polling: string | number, timeout: number, ...args: unknown[]) {
+  constructor(domWorld: DOMWorld, predicateBody: Function | string, predicateQueryHandlerBody: Function | string | undefined, title: string, polling: string | number, timeout: number, ...args: unknown[]) {
     if (helper.isString(polling))
       assert(polling === 'raf' || polling === 'mutation', 'Unknown polling option: ' + polling);
     else if (helper.isNumber(polling))
       assert(polling > 0, 'Cannot poll with non-positive interval: ' + polling);
     else
       throw new Error('Unknown polling options: ' + polling);
 
+    const wrapper =`
+    return (function wrapper(args) {
+      ${ predicateQueryHandlerBody ? `const predicateQueryHandler = ${predicateQueryHandlerBody};` : ''}
+      return (${predicateBody})(...args);
+    })(args)`;
+
     this._domWorld = domWorld;
     this._polling = polling;
     this._timeout = timeout;
-    this._predicateBody = helper.isString(predicateBody) ? 'return (' + predicateBody + ')' : 'return (' + predicateBody + ')(...args)';
+    this._predicateBody = helper.isString(predicateBody) ? 'return (' + predicateBody + ')' : wrapper;
```

Comment:
This is getting complicated enough to move to its own helper like

```suggestion
    this._predicateBody = getPredicateBody(predicateBody, predicateQueryHandlerBody);
```

which then does

```js
if (helper.isString(predicateBody)
  return `return (${predicateBody});`;
if (predicateQueryHandlerBody) {
  return `
    return (function wrapper(args) {
      const predicateQueryHandler = ${predicateQueryHandlerBody};
      return (${predicateBody})(...args);
    })(args);`;
}
return `return (${predicateBody})(...args);`;
```